io.zeebe.gateway.broker.BrokerClientTest is flaky

**Description**

```
[ERROR] Tests run: 12, Failures: 1, Errors: 0, Skipped: 0, Time elapsed: 35.462 s <<< FAILURE! - in io.zeebe.gateway.broker.BrokerClientTest

[ERROR] shouldEstablishNewConnectionsAfterDisconnect(io.zeebe.gateway.broker.BrokerClientTest)  Time elapsed: 9.73 s  <<< FAILURE!

java.lang.AssertionError: 

[List check last element] 

Expecting:

 <CONNECTED>

and actual:

 <CLOSED>

to refer to the same object

	at io.zeebe.gateway.broker.BrokerClientTest.shouldEstablishNewConnectionsAfterDisconnect(BrokerClientTest.java:198)

	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)

	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)

	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)

	at java.base/java.lang.reflect.Method.invoke(Method.java:566)

	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)

	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)

	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)

	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)

	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)

	at org.junit.internal.runners.statements.RunAfters.evaluate(RunAfters.java:27)

	at org.junit.rules.ExternalResource$1.evaluate(ExternalResource.java:48)

	at org.junit.rules.ExternalResource$1.evaluate(ExternalResource.java:48)

	at org.junit.rules.ExpectedException$ExpectedExceptionStatement.evaluate(ExpectedException.java:239)

	at org.junit.rules.TestWatcher$1.evaluate(TestWatcher.java:55)

	at org.junit.rules.RunRules.evaluate(RunRules.java:20)

	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:325)

	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:78)

	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:57)

	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)

	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)

	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)

	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)

	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)

	at org.junit.runners.ParentRunner.run(ParentRunner.java:363)

	at org.junit.runners.Suite.runChild(Suite.java:128)

	at org.junit.runners.Suite.runChild(Suite.java:27)

	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)

	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)

	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)

	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)

	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)

	at org.junit.runners.ParentRunner.run(ParentRunner.java:363)

	at org.apache.maven.surefire.junitcore.JUnitCore.run(JUnitCore.java:55)

	at org.apache.maven.surefire.junitcore.JUnitCoreWrapper.createRequestAndRun(JUnitCoreWrapper.java:137)

	at org.apache.maven.surefire.junitcore.JUnitCoreWrapper.executeLazy(JUnitCoreWrapper.java:119)

	at org.apache.maven.surefire.junitcore.JUnitCoreWrapper.execute(JUnitCoreWrapper.java:87)

	at org.apache.maven.surefire.junitcore.JUnitCoreWrapper.execute(JUnitCoreWrapper.java:75)

	at org.apache.maven.surefire.junitcore.JUnitCoreProvider.invoke(JUnitCoreProvider.java:158)

	at org.apache.maven.surefire.booter.ForkedBooter.invokeProviderInSameClassLoader(ForkedBooter.java:384)

	at org.apache.maven.surefire.booter.ForkedBooter.runSuitesInProcess(ForkedBooter.java:345)

	at org.apache.maven.surefire.booter.ForkedBooter.execute(ForkedBooter.java:126)

	at org.apache.maven.surefire.booter.ForkedBooter.main(ForkedBooter.java:418)



[INFO] 

[INFO] Results:

[INFO] 

[ERROR] Failures: 

[ERROR]   BrokerClientTest.shouldEstablishNewConnectionsAfterDisconnect:198 [List check last element] 

Expecting:

 <CONNECTED>

and actual:

 <CLOSED>

to refer to the same object
```

--- a/gateway/src/test/java/io/zeebe/gateway/broker/BrokerClientTest.java
+++ b/gateway/src/test/java/io/zeebe/gateway/broker/BrokerClientTest.java
@@ -51,8 +51,10 @@ import io.zeebe.util.sched.clock.ControlledActorClock;
 import io.zeebe.util.sched.future.ActorFuture;
 import java.util.ArrayList;
 import java.util.List;
-import java.util.concurrent.CopyOnWriteArrayList;
+import java.util.concurrent.CountDownLatch;
 import java.util.concurrent.ExecutionException;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicReference;
 import org.agrona.DirectBuffer;
 import org.junit.After;
 import org.junit.Before;
@@ -178,12 +180,23 @@ public class BrokerClientTest {
   }
 
   @Test
-  public void shouldEstablishNewConnectionsAfterDisconnect() {
+  public void shouldEstablishNewConnectionsAfterDisconnect() throws Exception {
     // given
     final ClientTransport clientTransport = ((BrokerClientImpl) client).getTransport();
-    final LoggingChannelListener channelListener = new LoggingChannelListener();
-    clientTransport.registerChannelListener(channelListener).join();
-    awaitConnectedChannels(channelListener, 1);
+    final AtomicReference<CountDownLatch> latchRef = new AtomicReference<>(new CountDownLatch(1));
+    clientTransport
+        .registerChannelListener(
+            new TransportListener() {
+              @Override
+              public void onConnectionEstablished(RemoteAddress remoteAddress) {
+                latchRef.get().countDown();
+              }
+
+              @Override
+              public void onConnectionClosed(RemoteAddress remoteAddress) {}
+            })
+        .join();
+    latchRef.get().await(10, TimeUnit.SECONDS);
 
     // ensuring an open connection
     broker.onExecuteCommandRequest(ValueType.DEPLOYMENT, DeploymentIntent.CREATE).doNotRespond();
@@ -191,11 +204,11 @@ public class BrokerClientTest {
 
     // when
     broker.closeTransport();
+    latchRef.set(new CountDownLatch(1));
     broker.bindTransport();
 
     // then
-    awaitConnectedChannels(channelListener, 2); // listener invocation is asynchronous
-    assertThat(channelListener.connectionState).last().isSameAs(ConnectionState.CONNECTED);
+    latchRef.get().await(10, TimeUnit.SECONDS);
   }
 
   @Test
@@ -383,34 +396,5 @@ public class BrokerClientTest {
             .done();
 
     builder.register();
-  }
-
-  private void awaitConnectedChannels(final LoggingChannelListener listener, final int count) {
-    waitUntil(
-        () ->
-            listener.connectionState.stream()
-                    .filter(state -> state == ConnectionState.CONNECTED)
-                    .count()
-                == count);
-  }
-
-  protected static class LoggingChannelListener implements TransportListener {
-
-    List<ConnectionState> connectionState = new CopyOnWriteArrayList<>();
-
-    @Override
-    public void onConnectionEstablished(final RemoteAddress remoteAddress) {
-      connectionState.add(ConnectionState.CONNECTED);
-    }
-
-    @Override
-    public void onConnectionClosed(final RemoteAddress remoteAddress) {
-      connectionState.add(ConnectionState.CLOSED);
-    }
-  }
-
-  protected enum ConnectionState {
-    CONNECTED,
-    CLOSED
   }
 }

